File: app/src/main/java/com/epam/valkaryne/openmovieapp/view/adapter/MoviesAdapter.java
package com.epam.valkaryne.openmovieapp.view.adapter;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.paging.PagingDataAdapter;
import androidx.recyclerview.widget.RecyclerView;

import com.epam.valkaryne.openmovieapp.data.api.model.MovieDataModel;
import com.epam.valkaryne.openmovieapp.databinding.ListItemMovieBinding;

public class MoviesAdapter extends PagingDataAdapter<MovieDataModel, MoviesAdapter.MovieViewHolder> {

    public MoviesAdapter() {
        super(new MoviesDiffCallback());
    }

    @NonNull
    @Override
    public MovieViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new MovieViewHolder(
                ListItemMovieBinding.inflate(
                        LayoutInflater.from(parent.getContext()),
                        parent,
                        false
                )
        );
    }

    @Override
    public void onBindViewHolder(@NonNull MovieViewHolder holder, int position) {
        MovieDataModel movie = getItem(position);
        if (movie != null) {
            holder.bind(movie);
        }
    }

    public static final class MovieViewHolder extends RecyclerView.ViewHolder {

        private ListItemMovieBinding binding;

        public MovieViewHolder(ListItemMovieBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        public void bind(MovieDataModel item) {
            binding.setMovie(item);
            binding.executePendingBindings();
        }
    }
}

File: app/src/main/java/com/epam/valkaryne/openmovieapp/view/adapter/MoviesAdapter.kt
package com.epam.valkaryne.openmovieapp.view.adapter

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.paging.PagingDataAdapter
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.epam.valkaryne.openmovieapp.common.loadImageFromUrl
import com.epam.valkaryne.openmovieapp.data.api.model.MovieInfo
import com.epam.valkaryne.openmovieapp.databinding.ListItemMovieBinding

/**
 * Adapter for the list of movies
 */
class MoviesAdapter :
    PagingDataAdapter<MovieInfo, MoviesAdapter.MovieViewHolder>(MoviesDiffCallback()) {

    override fun onBindViewHolder(holder: MovieViewHolder, position: Int) {
        val movie = getItem(position)
        movie?.let {
            holder.bind(movie)
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MovieViewHolder {
        return MovieViewHolder(
            ListItemMovieBinding.inflate(
                LayoutInflater.from(parent.context), parent, false
            )
        )
    }

    class MovieViewHolder(private val binding: ListItemMovieBinding) :
        RecyclerView.ViewHolder(binding.root) {

        fun bind(item: MovieInfo) {
            binding.movieItemTitle.text = item.title
            binding.movieItemYear.text = item.year
            binding.movieItemImage.loadImageFromUrl(item.posterUrl)
        }
    }
}

private class MoviesDiffCallback : DiffUtil.ItemCallback<MovieInfo>() {

    override fun areItemsTheSame(oldItem: MovieInfo, newItem: MovieInfo): Boolean {
        return oldItem.imdbID == newItem.imdbID
    }

    override fun areContentsTheSame(oldItem: MovieInfo, newItem: MovieInfo): Boolean {
        return oldItem == newItem
    }
}